### PR TITLE
refact: user context is cloned when passed to Decide APIs in Optimizely Class

### PR DESCRIPTION
**Pull request by ozayr-zaviar**
Summary
-------
-  `copy()` added in user_context to clone user_context before passing it in `decide`, `decide_for_keys`and `decide_all` as an argument

Test plan
---------
- Unit test added

### CODE REVIEW

File: src/Optimizely/OptimizelyUserContext.php [hunk 1]
```diff
@@ -31,24 +31,29 @@ public function __construct(Optimizely $optimizelyClient, $userId, array $attrib
         $this->attributes = $attributes;
     }
 
+    public function clone()
```
Comment (oakbani) [inline src/Optimizely/OptimizelyUserContext.php:34]: let's keep this protected for now

File: src/Optimizely/OptimizelyUserContext.php [hunk 2]
```diff
@@ -31,24 +31,29 @@ public function __construct(Optimizely $optimizelyClient, $userId, array $attrib
...
+
     public function setAttribute($key, $value)
     {
         $this->attributes[$key] = $value;
     }
 
     public function decide($key, array $options = [])
     {
-        return $this->optimizelyClient->decide($this, $key, $options);
+        return $this->optimizelyClient->decide($this->clone(), $key, $options);
```
Comment (oakbani) [inline src/Optimizely/OptimizelyUserContext.php:46]: Can you add a unit test that verifies that the OptimizelyUserContext object received in decide response is not the same object on which the decide was called?

**Review by oakbani — CHANGES_REQUESTED** [review]
Looks good. Can you check why the build is only failing on PHP 5.6 and 5.6? Is clone a reserved keyword?

File: tests/OptimizelyUserContextTests.php [hunk 3]
```diff
@@ -137,6 +137,22 @@ public function testDecideCallsAndReturnsOptimizelyDecideAPI()
...
+    {
+        $userId = 'test_user';
+        $attributes = [ "browser" => "chrome"];
+    
+        $optlyObject = new Optimizely($this->datafile);
+
+        $optUserContext = new OptimizelyUserContext($optlyObject, $userId, $attributes);
+        $decision = $optUserContext->decide('test_feature', ['DISABLE_DECISION_EVENT', 'ENABLED_FLAGS_ONLY']);
+        $optUserContext->setAttribute("test_key", "test_value");
+        $this->assertNotEquals(
```
Comment (oakbani) [inline tests/OptimizelyUserContextTests.php:150]: Why don't you directly compare the user context objects?

**Review by oakbani — APPROVED** [review]
Looks good. Just a minor comment.

File: tests/OptimizelyUserContextTests.php [hunk 4]
```diff
@@ -137,6 +137,22 @@ public function testDecideCallsAndReturnsOptimizelyDecideAPI()
...
+            $optUserContext->getAttributes(),
+            $decision->getUserContext()->getAttributes()
+        );
```
Comment (jaeopt) [inline tests/OptimizelyUserContextTests.php:153]: Can we also check if the returned userContext equal to the original user/attributes.

**Review by jaeopt — CHANGES_REQUESTED** [review]
Looks good. A test change suggested.

File: src/Optimizely/OptimizelyUserContext.php [hunk 5]
```diff
@@ -31,24 +31,29 @@ public function __construct(Optimizely $optimizelyClient, $userId, array $attrib
         $this->attributes = $attributes;
     }
 
+    protected function replicate()
```
Comment (jaeopt) [inline src/Optimizely/OptimizelyUserContext.php:34]: rename to "copy" (or "clone")? Are they reserved?

**Comment by oakbani** [comment]
@jaeopt  rebased to master and addressed comments.

**Review by jaeopt — APPROVED** [review]
LGTM

**Pull request merged**